Better show for parameterized methods (issue #10794)

mauro3 opened the pull request:
This only prints actual function parameters:
`f{T}(a::Array, b::T)  -> f{T}(a::Array{*,*}, b::T)`

fixes #10794

mauro3 committed:
Better show for parameterized methods

This only prints actual function parameters:
f{T}(a::Array, b::T)  -> f{T}(a::Array{_,_}, b::T)

mauro3 commented:
Well, this needs a bit of work.  I'll get to it in a couple of hours.

timholy commented:
I swear I've occasionally seen something like `Array{_<:FloatingPoint,_}` printed, but I can't recall the circumstances. But if I'm not imaging things, it seems that it would be best to use the same character in both places.

mauro3 commented:
Yes, I pretty sure I've seen this too but grepping through Base did not show anything.  Anyway, I'll go with `_` then.

mauro3 commented:
Updated and passing tests on my machine.  Also had to include the following method defs because of the the new Tuples:
```
length{T<:Tuple}(t::Type{T}) = nfields(t)
endof{T<:Tuple}(t::Type{T}) = length(t)
size{T<:Tuple}(t::Type{T}, d) = d==1 ? length(t) : throw(ArgumentError("invalid tuple dimension $d"))
getindex{T<:Tuple}(t::Type{T}, i) = t.types[i]
```

JeffBezanson commented:
Those methods are highly controversial. `T[x]` constructs an array when `T` is a type, so types cannot be "indexed". If we want a nicer way to access types we might need to come up with another API.

mauro3 commented:
Ok, I will remove them. In that case. I'll write it in terms of `nfields` and `t.types[i]`.

stevengj commented:
I also need `t.types` in PyCall (I want to `map` over them, currently via `Tuple{map(foo, t.types)...}`); it would be nice to have something documented for this.

mauro3 commented:
Removed those methods and rebased.

However, regarding indexing, most of the array construction should not interfere with indexing. For instance
```
getindex{T<:Tuple}(t::Type{T}, i::TT) = t.types[i]
```
wouldn't conflict with construction as long as TT is not a tuple.  Or am I missing something?

stevengj commented:
We could make tuple types iterable even if they are not indexable...
